syncer/syncer.go

```diff
@@ -0,0 +1,402 @@
+// ForceSync will be ignored.
+func (s *Syncer) ForceSync(ctx context.Context) {
+	s.logger.WithContext(ctx).Debug("executing ForceSync")
+	go s.synchronize(ctx)
+}
+
+func (s *Syncer) isClosed() bool {
+	select {
+	case <-s.shutdownCtx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func (s *Syncer) getSyncState() syncState {
+	return (syncState)(atomic.LoadUint32((*uint32)(&s.syncState)))
+}
+
+func (s *Syncer) setSyncState(ctx context.Context, newState syncState) {
+	oldState := syncState(atomic.SwapUint32((*uint32)(&s.syncState), uint32(newState)))
+	if oldState != newState {
+		s.logger.WithContext(ctx).With().Info("sync state change",
+			log.String("from", oldState.String()),
+			log.String("to", newState.String()),
+			log.FieldNamed("current", s.ticker.GetCurrentLayer()),
+			log.FieldNamed("latest", s.mesh.LatestLayer()),
+			log.FieldNamed("validated", s.mesh.ProcessedLayer()))
+		events.ReportNodeStatusUpdate()
+		// notify subscribes
+		s.awaitSyncedMu.Lock()
+		defer s.awaitSyncedMu.Unlock()
+		if newState == synced {
+			close(s.awaitSyncedCh)
+		} else if oldState == synced {
+			s.awaitSyncedCh = make(chan struct{})
+		}
+	}
+}
+
+// startNewSyncProcess returns true if syncer is not currently synchronizing.
+func (s *Syncer) startNewSyncProcess() bool {
+	return atomic.CompareAndSwapUint32(&s.isBusy, 0, 1)
+}
+
+func (s *Syncer) endSyncProcess() {
+	atomic.StoreUint32(&s.isBusy, 0)
+}
+
+func (s *Syncer) setTargetSyncedLayer(ctx context.Context, layerID types.LayerID) {
+	newSyncLayer := uint64(layerID)
+	oldSyncLayer := atomic.SwapUint64((*uint64)(&s.targetSyncedLayer), newSyncLayer)
+	s.logger.WithContext(ctx).With().Info("target synced layer changed",
+		log.Uint64("from", oldSyncLayer),
+		log.Uint64("to", newSyncLayer),
+		log.FieldNamed("current", s.ticker.GetCurrentLayer()),
+		log.FieldNamed("latest", s.mesh.LatestLayer()),
+		log.FieldNamed("validated", s.mesh.ProcessedLayer()))
+}
+
+func (s *Syncer) getTargetSyncedLayer() types.LayerID {
+	return types.LayerID(atomic.LoadUint64((*uint64)(&s.targetSyncedLayer)))
+}
+
+func (s *Syncer) synchronize(ctx context.Context) bool {
+	logger := s.logger.WithContext(ctx)
+
+	if s.isClosed() {
+		logger.Warning("attempting to sync while shutting down")
+		return false
+	}
+
+	// at most one synchronize process can run at any time
+	if !s.startNewSyncProcess() {
+		logger.Info("sync is already running, giving up")
+		return false
+	}
+	// no need to worry about race condition for s.run. only one instance of synchronize can run at a time
+	s.run++
+	logger.Info("staring sync run #%v", s.run)
+
+	s.setStateBeforeSync(ctx)
+	// start a dedicated process for validation.
+	// do not use a unbuffered channel for vQueue. we don't want it to block if the receiver isn't ready. i.e.
+	// if validation for the last layer is still running
+	vQueue := make(chan *types.Layer, s.ticker.GetCurrentLayer())
+	vDone := make(chan struct{})
+	go s.startValidating(ctx, s.run, vQueue, vDone)
```

Comment:
this will leak go routines as each time `synchronize` is called a new routine will be created
Perhaps move this code to constructor

Reply:
i don't believe this will cause goroutine leak. the `defer` call below close `vQueue` and block on `vDone`. once `vQueue` is closed, the range for loop in startValidating() will terminate and exit.

doing it in the constructor is also a viable option. my preference is to keep each sync run independent for ease of logging/debugging. if you think using just one validation goroutine is a better design, i will move it into the constructor.

Reply:
I stand corrected, missed the `close(vQueue)`